expr/func_movingwindow.go

```diff
@@ -0,0 +1,148 @@
+package expr
+
+import (
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/grafana/metrictank/batch"
+	"github.com/grafana/metrictank/schema"
+
+	"github.com/grafana/metrictank/api/models"
+	"github.com/grafana/metrictank/consolidation"
+	"github.com/raintank/dur"
+)
+
+type FuncMovingWindow struct {
+	in           GraphiteFunc
+	windowSize   string
+	fn           string
+	xFilesFactor float64
+
+	shiftOffset uint32
+}
+
+// NewMovingWindowConstructor takes an agg string and returns a constructor function
+func NewMovingWindowConstructor(name string) func() GraphiteFunc {
+	return func() GraphiteFunc {
+		return &FuncMovingWindow{fn: name}
+	}
+}
+
+func NewMovingWindow() GraphiteFunc {
+	return &FuncMovingWindow{fn: "average", xFilesFactor: 0}
+}
+
+func (s *FuncMovingWindow) Signature() ([]Arg, []Arg) {
+	return []Arg{
+			ArgSeriesList{val: &s.in},
+			ArgString{key: "windowSize", val: &s.windowSize, validator: []Validator{IsSignedIntervalString}},
+			ArgString{key: "func", opt: true, val: &s.fn, validator: []Validator{IsConsolFunc}},
+			ArgFloat{key: "xFilesFactor", opt: true, val: &s.xFilesFactor, validator: []Validator{WithinZeroOneInclusiveInterval}},
+		}, []Arg{
+			ArgSeriesList{},
+		}
+}
+
+func (s *FuncMovingWindow) Context(context Context) Context {
+	var err error
+	s.shiftOffset, err = s.getWindowSeconds()
+	if err != nil {
+		// shouldn't happen, validated above
+		return context
+	}
+
+	// Adjust to fetch from the shifted context.from with
+	// context.to unchanged
+	context.from -= s.shiftOffset
+	return context
+}
+
+func (s *FuncMovingWindow) Exec(dataMap DataMap) ([]models.Series, error) {
+	series, err := s.in.Exec(dataMap)
+	if err != nil {
+		return nil, err
+	}
+
+	aggFunc := consolidation.GetAggFunc(consolidation.FromConsolidateBy(s.fn))
+	aggFuncName := fmt.Sprintf("moving%s", strings.Title(s.fn))
+	formatStr := fmt.Sprintf("%s(%%s,\"%s\")", aggFuncName, s.windowSize)
+
+	newName := func(oldName string) string {
+		return fmt.Sprintf(formatStr, oldName)
+	}
+
+	outputs := make([]models.Series, 0, len(series))
+	for _, serie := range series {
+
+		serie.Target = newName(serie.Target)
+		serie.QueryPatt = newName(serie.QueryPatt)
+		serie.Tags = serie.CopyTagsWith(aggFuncName, s.windowSize)
+		serie.QueryFrom, serie.Datapoints = aggregateOnWindow(serie, aggFunc, s.shiftOffset, s.xFilesFactor)
+
+		outputs = append(outputs, serie)
+	}
+
+	dataMap.Add(Req{}, outputs...)
+	return outputs, nil
+}
+
+func aggregateOnWindow(serie models.Series, aggFunc batch.AggFunc, interval uint32, xFilesFactor float64) (uint32, []schema.Point) {
+	out := pointSlicePool.Get().([]schema.Point)
+
+	numPoints := len(serie.Datapoints)
+	serieStart, serieEnd := serie.QueryFrom, serie.QueryTo
+	queryStart := uint32(serieStart + interval)
+
+	if queryStart < serieStart {
+		panic("Query start is before the first point in series")
```

Comment:
Not sure that a panic is the best way to handle this. I do believe that panics get caught at a higher level from within HTTP requests, but it sort of skips around the normal flow. Any reason not to just return an error?

Reply:
I saw some examples of panic calls in the code and thought that may be the preferred way. I'll investigate if return is the preferred way and will make the change accordingly.

Reply:
couldn't find an indication that weighs in either way. So i've made changes to return the error and play it safe